calico/felix/labels.py

```diff
@@ -0,0 +1,480 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from calico.calcollections import MultiDict
+from calico.felix.selectors import LabelToLiteralEquality, LabelInSetLiteral
+
+_log = logging.getLogger(__name__)
+
+
+class LinearScanLabelIndex(object):
+    """
+    A label index matches a set of SelectorExpressions against a set of
+    label dicts.  As the matches between the two collections change,
+    it triggers calls to on_match_started/on_match_stopped.
+
+    A simple implementation of a label index.  Every update is handled by
+    a full linear scan.
+
+    This class has two purposes:
+
+    - it provides a benchmark against which other implementations can be
+      measured
+    - since it's simple, it's useful for comparative testing; any
+      other label index implementation should give the same end result.
+    """
+
+    def __init__(self):
+        # Cache of raw label dicts by item_id.
+        self.labels_by_item_id = {}
+        # All expressions by ID.
+        self.expressions_by_id = {}
+        # Map from expression ID to matching item_ids.
+        self.matches_by_expr_id = MultiDict()
+        self.matches_by_item_id = MultiDict()
+        self.total_evaluations = 0
+
+    def on_expression_update(self, expr_id, expr):
+        """
+        Called to update a particular expression.
+
+        Triggers events for match changes.
+        :param expr_id: an opaque (hashable) ID to associate with the
+               expression.  There can only be one expression per ID.
+        :param expr: The SelectorExpression to add to the index or None to
+               remove it.
+        """
+        _log.debug("Expression %s updated to %s", expr_id, expr)
+        self._scan_all_labels(expr_id, expr)
+        self._store_expression(expr_id, expr)
+
+    def on_labels_update(self, item_id, new_labels):
```

Comment:
Agree it'd be cleaner but I'm a bit wary of trying to replace that many instances at this stage (there are hundreds of references to labels in the code and I don't want to introduce a rename bug).